src/adapter/src/coord/introspection.rs

```diff
@@ -0,0 +1,508 @@
+            move || {
+                span.in_scope(|| {
+                    // MIR ⇒ LIR lowering and LIR ⇒ LIR optimization (global)
+                    let global_lir_plan =
+                        optimizer.catch_unwind_optimize(global_mir_plan.clone())?;
+
+                    let stage = IntrospectionSubscribeStage::Finish(IntrospectionSubscribeFinish {
+                        validity,
+                        global_lir_plan,
+                        read_holds,
+                    });
+                    Ok(Box::new(stage))
+                })
+            },
+        )))
+    }
+
+    async fn sequence_introspection_subscribe_finish(
+        &mut self,
+        stage: IntrospectionSubscribeFinish,
+    ) -> Result<StageResult<Box<IntrospectionSubscribeStage>>, AdapterError> {
+        let IntrospectionSubscribeFinish {
+            validity,
+            global_lir_plan,
+            read_holds,
+        } = stage;
+
+        let cluster_id = validity.cluster_id.expect("always set");
+        let replica_id = validity.replica_id.expect("always set");
+        let subscribe_id = global_lir_plan.sink_id();
+
+        // The subscribe may already have been dropped, in which case we must not install a
+        // dataflow for it.
+        if self.introspection_subscribes.contains_key(&subscribe_id) {
+            let (df_desc, _df_meta) = global_lir_plan.unapply();
+            self.ship_dataflow(df_desc, cluster_id).await;
+
+            self.controller
+                .compute
+                .set_subscribe_target_replica(cluster_id, subscribe_id, replica_id)
+                .expect("cannot fail");
+        }
+
+        drop(read_holds);
+
+        let (_tx, rx) = mpsc::unbounded_channel();
+        let resp = ExecuteResponse::Subscribing {
+            rx,
+            ctx_extra: ExecuteContextExtra::default(),
+            instance_id: cluster_id,
+        };
+        Ok(StageResult::Response(resp))
+    }
+
+    /// Drops the introspection subscribes installed on the given replica.
+    ///
+    /// Dropping an introspection subscribe entails:
+    ///  * removing it from [`Coordinator::introspection_subscribes`]
+    ///  * dropping its compute collection
+    ///  * retracting any rows previously omitted by it from its corresponding storage-managed
+    ///    collection
+    pub(super) fn drop_introspection_subscribes(&mut self, replica_id: ReplicaId) {
+        let to_drop: Vec<_> = self
+            .introspection_subscribes
+            .iter()
+            .filter(|(_, s)| s.replica_id == replica_id)
+            .map(|(id, _)| *id)
+            .collect();
+
+        for id in to_drop {
+            self.drop_introspection_subscribe(id);
+        }
+    }
+
+    fn drop_introspection_subscribe(&mut self, id: GlobalId) {
+        let Some(subscribe) = self.introspection_subscribes.remove(&id) else {
+            soft_panic_or_log!("attempt to remove unknown introspection subscribe (id={id})");
```

Comment:
I have become very careful of panicking in envd. In the past it happened a couple times that we got into a crashloop due to some bug in the code, and then could resolve this only be pushing a patch that changed the panicking assert to a soft one. Even if we don't crashloop, a panic now and then might make the environment unstable, causing a maybe harmless bug in the code to be elevated to a high-urgency one. There are of course still asserts that need to be hard because they protect correctness.

In this case I don't see how correctness could be impacted. Most likely we get here by trying to drop an introspection subscribe twice, in which case ignoring the second drop is probably what we want. Less likely, we have somehow mixed up subscribe IDs and forget to drop the subscribe we really wanted to drop, in which case some coordinator memory will be wasted and some entries will not be cleaned up from an introspection relation until the next restart. This seems preferable to panicking for me.